core/src/main/java/jenkins/agents/CloudSet.java

```diff
@@ -0,0 +1,219 @@
+        jenkins.checkPermission(Jenkins.ADMINISTER);
+
+        if (mode != null && mode.equals("copy")) {
+            name = checkName(name);
+
+            Cloud src = jenkins.getCloud(from);
+            if (src == null) {
+                if (Util.fixEmpty(from) == null) {
+                    throw new Failure(Messages.CloudSet_SpecifyCloudToCopy());
+                } else {
+                    throw new Failure(Messages.CloudSet_NoSuchCloud(from));
+                }
+            }
+
+            // copy through XStream
+            String xml = Jenkins.XSTREAM.toXML(src);
+            // Not great, but cloud name is final
+            xml = xml.replace("<name>" + src.name + "</name>", "<name>" + name + "</name>");
+            Cloud result = (Cloud) Jenkins.XSTREAM.fromXML(xml);
+            jenkins.clouds.add(result);
+            // send the browser to the config page
+            rsp.sendRedirect2(result.getUrl() + "configure");
+        } else {
+            // proceed to step 2
+            if (mode == null) {
+                throw new Failure("No mode given");
+            }
+
+            Descriptor<Cloud> d = Cloud.all().findByName(mode);
+            if (d == null) {
+                throw new Failure("No node type ‘" + mode + "’ is known");
+            }
+            handleNewCloudPage(d, name, req, rsp);
+        }
+    }
+
+    private void handleNewCloudPage(Descriptor<Cloud> descriptor, String name, StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, Descriptor.FormException {
+        checkName(name);
+        JSONObject formData = req.getSubmittedForm();
+        formData.put("name", name);
+        formData.put("cloudName", name);
+        formData.remove("mode"); // Cloud descriptors won't have this field.
+        Cloud instance = descriptor.newInstance(req, formData); // Not great but that's the best I have so far to pass the given name.
+        req.setAttribute("instance", instance);
+        req.setAttribute("descriptor", descriptor);
+        req.getView(this, "_new.jelly").forward(req, rsp);
+    }
+
+    /**
+     * Really creates a new agent.
+     */
+    @POST
+    public synchronized void doDoCreateItem(StaplerRequest req, StaplerResponse rsp,
+                                            @QueryParameter String type) throws IOException, ServletException, Descriptor.FormException {
+        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+        Jenkins.get().clouds.add(Cloud.all().find(type).newInstance(req, req.getSubmittedForm()));
+        // take the user back to the cloud list top page
+        rsp.sendRedirect2(".");
+    }
+
+    @Extension
+    public static class DescriptorImpl extends Descriptor<CloudSet> {
+        /**
+         * Auto-completion for the "copy from" field in the new cloud page.
+         */
+        @SuppressWarnings("unused") // stapler
+        public AutoCompletionCandidates doAutoCompleteCopyNewItemFrom(@QueryParameter final String value) {
```

Comment:
Weird naming scheme with `…Item`, given that clouds aren't `Item`s.

Reply:
Due to https://github.com/jenkinsci/jenkins/blob/5870838f690eaa06762630eceb21c40d3df0ceb8/core/src/main/resources/lib/hudson/newFromList/form.jelly#L93. Could possibly be addressed separately.